### ISSUE TITLE
docs: automate listing all stock/built-in types for typed extensions

### ISSUE THREAD

**Issue by mattklein123** [envoyproxy/envoy#13167]
A major problem with our documentation is the user does not know what types are available for all of the typed extension points. Obviously we can't know all custom extensions and all bets are off if the user is doing a custom build, but we can ignore those cases because those users won't have any issues.

For everyone else, I think we should:
1) Dump out all extensions in the stock build (either do this with bazel ahead of time, or somehow dump it out of the final exe). This is a map of extension type/category/link to docs, etc.
2) Automate the documentation so all typed_extensions are printed with docs listing all stock options, as well as links to relevant docs, as well as sample config.

**Comment by mattklein123** [comment]
Cross linking to https://github.com/envoyproxy/envoy/issues/13078 as some variant of this is going to be needed to make the vscode plugin work correctly.

**Comment by phlax** [comment]
part 1 of this is resolved in #14721 

the second part:

> Automate the documentation so all typed_extensions are printed with docs listing all stock types, as well as links to relevant docs, as well as sample config.

im wondering if we need to do something further in terms of sample config

i guess one extreme would be for somehow *every* extension to programatically provide a sample config that can be embedded in the extension category docs - so pretty much all extensions are shown in some form on the extension category page.

i think this risks creating quite a bit of duplication and on a page with a lot of known extensions would make  the page pretty huge

on the other hand, we could use tabs or a dropdown or somesuch to switch between the examples so it might work

the advantage of this approach is that it minimizes the redirection required to get the information, its in one place and you can click through for more info, rather than just signposted from one place, and clicking back <> forward to figure out *your* config

one thing im also conscious of here is that some of the existing protos have manually listed extensions/info etc which is now kinda duplicated by the links that we added. Probably best to just remove these i guess - ill create a cleanup PR, and we can discuss thiss there

**Comment by phlax** [comment]
@mattklein123 @htuch ^^

**Comment by mattklein123** [comment]
> i guess one extreme would be for somehow every extension to programatically provide a sample config that can be embedded in the extension category docs - so pretty much all extensions are shown in some form on the extension category page.

In a perfect world, yes, I think we should require sample configs for all extensions, and given we can type_check doc examples, we would require that all examples use the type checked doc plugin so it never gets out of date. In an interim world, I wonder if we could start with a snipet that just shows the type_url/@type and an empty config stanza with some '...' or something. I'm not sure what is best here but I would rather make incremental progress vs. wait until we make everything perfect. Will defer to you though on what you think is best.

> one thing im also conscious of here is that some of the existing protos have manually listed extensions/info etc which is now kinda duplicated by the links that we added. Probably best to just remove these i guess - ill create a cleanup PR, and we can discuss thiss there

+1

**Comment by htuch** [comment]
Why not use the additional extension metadata YAML in https://github.com/envoyproxy/envoy/issues/14470? We could mandate all new extensions populate a field in the YAML that provides an Any configuration. Tooling can type check this in CI and render it to docs as suggested by @phlax. We could backfill existing extensions over time (create a beginner task in issues). WDYT?

**Comment by mattklein123** [comment]
Yeah as long as it is type checked and required, any mechanism is fine with me. This will be amazing!
